### DISCUSSION TITLE
ExecuteAsync with ResilienceContext and CancellationToken

### DISCUSSION THREAD

**Discussion by AndrejNazarov** [App-vNext/Polly#2133] [Q&A] [answered]
Hello,

I wasn't able to find an `ExecuteAsync` method on `ResiliencePipeline` that allows to pass a `ResilienceContext` and a `CancellationToken` together.

This can be done with AsyncPolicy.

My use-case: 
- retry an async operation that should be cancellable during retries, thus passing a CancellationToken from the caller
- working with `Context` (newly `ResilienceContext`) to pass specific `ILogger` into the retry code and some other properties
- my pipeline definition (i.e. `OnRetry` method) is in a different class than the decorated method, I want to inject it via DI (either via `ResiliencePipelineRegistry<string>` or `ResiliencePipelineProvider<string>`)

I'm unable to do this in Polly v8 - the `ExecuteAsync` methods on `ResiliencePipeline` class contain either a `CancellationToken` or a `ResilienceContext`, but never both.

Here is some sample code

```cs
public class MyService(ILogger logger, 
    ResiliencePipelineProvider<string> pipelineProvider,
    AsyncPolicy asyncPolicy,
    ISomethingAsync foo) : BackgroundService 
{
    protected override async Task ExecuteAsync(CancellationToken stoppingToken) {
        var pipeline = pipelineProvider.GetPipeline("MyPipeline");
        while (!stoppingToken.IsCancellationRequested)
        {
            //new way
            var resilienceContext = ResilienceContextPool.Shared.Get(stoppingToken);
            resilienceContext.Properties.Set(new ResiliencePropertyKey<ILogger>("Logger"), logger);
            try
            {
                //something like this, but this method signature doesn't exist
                var resultNew = await pipeline.ExecuteAsync(async (ctx, token) =>
                {
                    ctx.Properties.Set(new ResiliencePropertyKey<string>("SomeKey"), "SomeValue");
                    var bar = await foo.GetAsync(token);
                    return bar;
                }, resilienceContext, stoppingToken);
            }
            finally
            {
                ResilienceContextPool.Shared.Return(resilienceContext);
            }
            
            //pre version 8 - works
            var context = new Context();
            context.Add("Logger", logger);
            var resultOld = await asyncPolicy.ExecuteAsync(async (ctx, token) =>
            {
                ctx["SomeKey"] = "SomeValue";
                var bar = await foo.GetAsync(token);
                return bar;
            }, context, stoppingToken);
        }
    }
}
```
Hope this makes sense.

**Comment by martincostello** [comment]
The reason there aren't any overloads where both are available is because the cancellation token is a member of the resilience context:

https://github.com/App-vNext/Polly/blob/0856394583acee9237c8189347a06d83d362f479/src/Polly.Core/ResilienceContext.cs#L34

**Reply by AndrejNazarov** [reply]
I see. Can you then suggest how to do this with ResilienceContext? If I do it this way, the token inside the lambda is different:
```cs
while (!stoppingToken.IsCancellationRequested)
{
    Console.WriteLine($"stoppingToken wait handle:{stoppingToken.WaitHandle.SafeWaitHandle.GetHashCode()}");
    
    //is this how I pass the cancellation token to the lambda?
    var resilienceContext = ResilienceContextPool.Shared.Get(stoppingToken);
    resilienceContext.Properties.Set(new ResiliencePropertyKey<ILogger<MyService>>("Logger"), logger);
    try
    {
        var resultNew = await pipeline.ExecuteAsync(async (ctx) =>
        {
            ctx.Properties.Set(new ResiliencePropertyKey<string>("SomeKey"), "SomeValue");
            var bar = await foo.GetAsync(ctx.CancellationToken);
            //this returns a different HashCode than stoppingToken above, so it's a different CancellationToken
            Console.WriteLine($"Lambda token wait handle:{ctx.CancellationToken.WaitHandle.SafeWaitHandle.GetHashCode()}"); 
            return bar;
        }, resilienceContext);
    }
    finally
    {
        ResilienceContextPool.Shared.Return(resilienceContext);
    }
}
```
If I do it the old way, the token inside the lambda is the same (as you would expect).
```cs
    var polly = Policy.Handle<Exception>().RetryAsync(3, (exception, retryCount, context) => Console.WriteLine($"try: {retryCount}, Exception: {exception.Message}"));
    var context = new Context();
    var resultOld = await polly.ExecuteAsync(async (ctx, token) =>{
       ctx["SomeKey"] = "SomeValue";
       var bar = await foo.GetAsync(token);
       //This returns same hash code as stoppingToken, so it is the same token
       Console.WriteLine($"Lambda token wait handle :{token.WaitHandle.SafeWaitHandle.GetHashCode()}");
       return bar;
}, context, stoppingToken);
```

What  am I doing wrong?

**Reply by martincostello** [reply]
The token you get on the context will be different because it's internally wired up to honour timeouts related to a timeout policy/strategy etc. That's an implementation detail you shouldn't worry about - you should just use the token you are given to honour cancellations. If you want to use a _specific_ token, then you should pass it into the context with a named value and use it directly like you are with other things.

**Reply by AndrejNazarov** [reply]
Okay. Let's say I have an async method that takes 30s to complete. 
- I want it to be cancellable from outside
- I want it to run inside a Polly ExecuteAsync wrapper for retries and all the other Polly goodness
- I want to use ResilienceContext data for e.g. specific loggers, etc

What's the best way to achieve this?

This seems to pass/reference the same token:
```cs
var resilienceContext = ResilienceContextPool.Shared.Get(outsideToken); // is it necessary to pass the outsideToken here?
ResiliencePropertyKey<CancellationToken> TokenPropertyKey = new("TokenKey");
ResiliencePropertyKey<string> StringPropertyKey = new("SomeKey");
resilienceContext.Properties.Set(TokenPropertyKey, outsideToken);
try
{
    var resultNew = await pipeline.ExecuteAsync(async (ctx) =>
    {
        ctx.Properties.Set(StringPropertyKey, "SomeValue");
        var token = ctx.Properties.GetValue(TokenPropertyKey, default);
        var bar = await foo.GetAsync(token);
        Debug.Assert(outsideToken.WaitHandle.SafeWaitHandle.GetHashCode() == token.WaitHandle.SafeWaitHandle.GetHashCode(), "Tokens are different");
        return bar;
    }, resilienceContext);
}
finally
{
    ResilienceContextPool.Shared.Return(resilienceContext);
}
```

But this looks rather complicated compared to what was before. Is there no other/simpler way?

**Reply by martincostello** [reply, marked as answer]
Polly v7 had far too many method overloads which made future iteration and maintenance difficult, so we rationalised down the number of overloads for v8. We also made other changes to improve performance, such as pooling contexts.

If you want to use both features (context _and_ cancellation) then the shared snippet is the way to achieve this. You could simplify the code to some degree by adding helper methods specific to your context to do things like get/set your values on your `ResilienceContext` via methods that abstract away the resilience keys etc: [example](https://github.com/martincostello/polly-sandbox/blob/4766ec17350253738edbd9fc96e2626d2624a88e/src/PollySandbox/Extensions/PollyExtensions.cs#L11-L37).

The inner token in the lambda will/should be different if you're using strategies like timeout and hedging, so I would avoid writing your code to assume that the token is the same as your outside token, then you don't need to worry about adding and getting it to the context.

**Comment by AndrejNazarov** [comment]
> The inner token in the lambda will/should be different if you're using strategies like timeout and hedging,

Thank you for this (and other) clarifications. I wasn't aware of that. I ran it through the debugger and indeed these 2 tokens are different in v8:
```cs
var res = await pipeline.ExecuteAsync(async (innerToken) =>
{
    var bar = await foo.GetAsync(innerToken);
    return bar;
}, outsideToken);
```
but are the same in v7 equivalent code.
